kernel/debug: add missing safety comments

stefano-garzarella opened the pull request:
Add safety comments around `unsafe` blocks and mark some functions `unsafe` if the safety is demanded to the caller.
Add a FIXME in place where unsafe could be very dangerous and silence `clippy::undocumented_unsafe_blocks` lint in gdbstub.rs
Mark all 'gdbstub' public functions unsafe.
Add a compilation error if gdbstub is enabled in release (we already had a comment on discourage that)

stefano-garzarella commented on kernel/src/debug/gdbstub.rs:
```diff
@@ -477,6 +487,7 @@ pub mod svsm_gdbstub {
                     // The registers are stored in the top of the task stack as part of the
                     // saved context. We need to switch to the task pagetable to access them.
                     let _task_context = GdbTaskContext::switch_to_task(tid.get() as u32);
+                    // SAFETY: TODO
```
Should we do some checks on `task.rsp` or can we trust the value as it is?

@roy-hopkins @joergroedel any clue on this?

joergroedel replied:
In general there is likely a better way to implement accesses to a foreign task context. I need to think about this some more.

stefano-garzarella commented on kernel/src/debug/stacktrace.rs:
```diff
@@ -134,8 +136,10 @@ impl StackUnwinder {
+        // SAFETY: TODO
         let rbp = unsafe { rsp.as_ptr::<VirtAddr>().read_unaligned() };
         let rsp = rsp + mem::size_of::<VirtAddr>();
+        // SAFETY: TODO
```
Not sure if we should mark the entire function unsafe and demand to the caller

joergroedel replied:
This function always works on the stacks of the current context, so de-referencing pointers from the stacks of the context is safe.

stefano-garzarella commented:
I need some help on some safety comments :-)

joergroedel commented on kernel/src/debug/gdbstub.rs:
```diff
@@ -196,6 +204,7 @@ pub mod svsm_gdbstub {
                 let cr3 = read_cr3();
                 let task = tl.get_task(id);
                 if let Some(task) = task {
+                    // SAFETY: TODO
```
@roy-hopkins might know better, but I think this is safe because all of GDBStub and the memory it uses are globally mapped and are thus accessible from any task context.

joergroedel commented on kernel/src/debug/gdbstub.rs:
```diff
@@ -211,6 +220,7 @@ pub mod svsm_gdbstub {
     impl Drop for GdbTaskContext {
         fn drop(&mut self) {
             if self.cr3 != 0 {
+                // SAFETY: TODO
```
This is safe because it just restores the previous page-table, given that the percpu mappings is taken care of.

stefano-garzarella commented:
As discussed in TSC meeting, debug code is really unsafe in some places, so the idea is to silence `clippy` in these files and make public API `unsafe`, so that we will use it carefully.

stefano-garzarella commented:
v2:
- removed all `SAFETY: TODO`
- added some`SAFETY: `  comments provided by @joergroedel 
- added a `FIXME` in place where `unsafe` could be very dangerous
- marked all public functions unsafe [@msft-jlange]
- added a compilation error if gdbstub is enabled in release (we already had a comment on discourage that)
- silenced clippy::undocumented_unsafe_blocks lint in gdbstub.rs [@msft-jlange]

@joergroedel @msft-jlange PTAL!

stefano-garzarella commented:
v3:
- fixed wrong SAFETY comment in `kernel/src/cpu/idt/svsm.rs`

msft-jlange commented on kernel/src/debug/gdbstub.rs:
```diff
@@ -4,13 +4,24 @@
 //
 #[cfg(feature = "enable-gdb")]
 pub mod svsm_gdbstub {
+    // Comment the following 2 lines if you really want to enable gdbstub in
+    // release build because memory safety is not guaranteed.
```
Excluding debugger support from release builds is about security, not about memory safety.  Including the `compile_error` is a good idea just to create a higher barrier to prevent accidental inclusion, but we should still be accurate with the comments.

stefano-garzarella replied:
Okay, I'll update the comment. thanks.

msft-jlange commented on kernel/src/debug/gdbstub.rs:
```diff
@@ -81,7 +102,12 @@ pub mod svsm_gdbstub {
+    /// Some gdbstub operations are very dangerous and could compromise memory
+    /// safety, but these are not avoidable for now, so the caller must be sure
+    /// to use gdbstub carefully (and only in debug mode).
+    pub unsafe fn handle_debug_exception(ctx: &mut X86ExceptionContext, exception: usize) {
```
It's not clear to me that this function is really unsafe, because it does not interpret the caller's parameters in a way that could affect memory safety.  The function already appears to include sufficient logic to prevent execution outside of debugger support, and debugger support isn't expected to maintain safety anyway.

stefano-garzarella replied:
Perhaps I misunderstood when we discussed this in the TSC. But I thought we agreed to mark all public functions of the module as unsafe because, in some cases, we cannot guarantee safety internally.

If I get it wrong, I can revert this change.

msft-jlange commented on kernel/src/debug/gdbstub.rs:
```diff
@@ -44,11 +55,21 @@ pub mod svsm_gdbstub {
     // The static mutable reference to the stack is protected by the GDB_STATE lock.
     #[allow(static_mut_refs)]
-    pub fn gdbstub_start(platform: &'static dyn SvsmPlatform) -> Result<(), u64> {
+    pub unsafe fn gdbstub_start(platform: &'static dyn SvsmPlatform) -> Result<(), u64> {
```
I'm not sure this function really needs to be `unsafe`.  Enabling the debugger itself does not compromise memory safety; it is use of the debugger that does.  Code should only be marked `unsafe` if the code itself is unsafe, not just because it enables a scenario that could compromise safety later.

stefano-garzarella replied:
Same of https://github.com/coconut-svsm/svsm/pull/744/files#r2235121532

msft-jlange commented on kernel/src/debug/gdbstub.rs:
```diff
@@ -752,11 +795,29 @@ pub mod svsm_gdbstub {
+    /// Starting the gdbstub is not multi-thread safe, so the caller must
+    /// ensure to call this function on a single processor.
+    ///
+    /// Some gdbstub operations are very dangerous and could compromise memory
+    /// safety, but these are not avoidable for now, so the caller must be sure
+    /// to use gdbstub carefully (and only in debug mode).
```
And if you remove the `unsafe` declaration from the gdb-enabled version, then obviously this whole comment can go away.

stefano-garzarella replied:
Good point (for both), let's discuss on https://github.com/coconut-svsm/svsm/pull/744/files#r2235121532 next step and I'd fix in a way or the other.

stefano-garzarella commented on kernel/src/debug/gdbstub.rs:
```diff
@@ -164,11 +190,18 @@ pub mod svsm_gdbstub {
+    /// Some gdbstub operations are very dangerous and could compromise memory
+    /// safety, but these are not avoidable for now, so the caller must be sure
+    /// to use gdbstub carefully (and only in debug mode).
+    pub unsafe fn debug_break() {
```
Same of https://github.com/coconut-svsm/svsm/pull/744#discussion_r2235121532

stefano-garzarella commented on kernel/src/debug/gdbstub.rs:
```diff
@@ -187,15 +220,21 @@ pub mod svsm_gdbstub {
+        // a better way since this is a debug module.
         #[must_use = "The task switch will have no effect if the context is dropped"]
-        fn switch_to_task(id: u32) -> Self {
+        unsafe fn switch_to_task(id: u32) -> Self {
```
Okay, but this is one of the function that it's really unsafe IIUC, what comment should we write?

stefano-garzarella replied:
I just noticed that I added `FIXME` to explain why this is really unsafe, so not sure what to do here, since the user of this function can't provide any guarantee IIUC.

msft-jlange commented on kernel/src/debug/stacktrace.rs:
```diff
@@ -55,6 +55,8 @@ fn is_stage2() -> bool {
 impl StackUnwinder {
     pub fn unwind_this_cpu() -> Self {
         let mut rbp: usize;
+        // SAFETY: Inline assembly to read RPB, which does not change any state
```
```suggestion
// SAFETY: Inline assembly to read RBP, which does not change any state
```

msft-jlange commented:
@stefano-garzarella I think this is moot now that #770 has been merged.  Do you agree?
